accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +268,157 @@ The **flatten** function is modified as follows:
+with declared return type `Future<Future<int>>`. Conversely, it is no longer
+allowed to return a `Future<dynamic>` or `FutureOr<dynamic>` when the future
+value type is `Future<U>` for some `U` which is not a top type, because it is
+no longer enough that `dynamic` is assignable to `U`._
+
+The dynamic semantics specified at
+[this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15597)
+is changed as follows, where `$f$` is the enclosing function with declared
+return type `$T$`, and `$e$` is the returned expression with static type `$S$`:
+
+```
+When $f$ is a synchronous non-generator, evaluation proceeds as follows:
+The expression $e$ is evaluated to an object $o$.
+A dynamic error occurs unless the dynamic type of $o$ is a subtype of
+the actual return type of $f$
+(\ref{actualTypes}).
+Then the return statement $s$ completes returning $o$
+(\ref{statementCompletion}).
+
+\commentary{%
+The case where the evaluation of $e$ throws is covered by
+the general rule which propagates the throwing completion,
+in this case to the call site.%
+}
+
+When $f$ is an asynchronous non-generator with future value type $T_v$
+(\ref{functions}),
+the return statement $s$ is desugared at compile time as described below.
+The resulting code uses \code{\_return $e'$},
+which is a statement that evaluates $e'$ to an object
+$o'$, completes the returned future of $f$ with $o'$,
+and finally returns without an object
+(\ref{statementCompletion}),
+thus terminating the execution of $f$.
+Alternatively, if the evaluation of $e'$ throws
+an object \metavar{error} and stack trace \metavar{stackTrace},
+the returned future of $f$ is error-completed
+with \metavar{error} and \metavar{stackTrace},
+and finally the statement returns without an object.
+The desugaring proceeds as follows,
+where $\Gamma$ is the typing environment of $s$,
+and the first item whose requirement is satisfied is used:
+
+\begin{itemize}
+\item% 1
+  When $S$ is \DYNAMIC:
+  \code{\VAR r = $e$; \_return r \IS{} Future<$T_v$> ? \AWAIT{} r : (r \AS{} $T_v$);}
+\item% 2
+  When \SubtypeStd{S}{T_v}: \code{\_return $e$;}
+\item% 3
+  When \SubtypeStd{S}{\code{Future<$T_v$>}}: \code{\_return \AWAIT{} $e$;}
+\item% 4
+  When $S$ is of the form \code{FutureOr<$U$>} where \SubtypeStd{U}{T_v}:
+  \code{\VAR r = $e$; \_return r is Future<$T_v$> ? \AWAIT{} r : r;}
+\end{itemize}
+
+\commentary{%
+Note that one of the above cases apply, or a compile-time error has occurred.%
+% It is required at compile-time that S is assignable to T_v, or that
+% flatten(S) <: T_v. If S is assignable to T_v then item 1 or 2 will apply.
+% Otherwise flatten(S) <: T_v. We know that flatten(S) != S, or S would be
+% assignable to T_v (and item 1 or 2 would apply). So S is of the form
+% FutureOr<S1> where S1 <: T_v; or S implements Future<S1> and
+% flatten(S) == S1. In the latter case item 3 applies, and in the former case
```

Comment:
We do not know that those are the only forms (any more). As mentioned above, we need to account for `?` types.

So, let's be extra careful about this case analysis.

It's still *correct* to simply have a 3rd item of the form:

  3. If *U* = *flatten*(*S*), *U* <: *T<sub>v</sub>*: `var r = e; return r is Future<U> ? await r : r;`.

The non-await branch might be unreachable if *S* is a non-FutureOr type, but that can just be an implementation optimization.

Reply:
I changed item 2 to cover the cases, PTAL there.